### DISCUSSION TITLE
command line: parallelization on cluster

### DISCUSSION THREAD

**Discussion by JM-Bader** [vdemichev/DiaNN#379] [Q&A] [answered]
I would like to process >1000 diaPASEF runs on a linux computer cluster with 56 nodes and 24 cpu per node. 
Running command line DIA-NN on 1 node with 1 core works fine. Now I would like to speed the processing up. 
The github only provides the --threads command, but surprisingly no "--node" command. 

On the computer cluster, I specify nodes, tasks per node, and cpus per node with a scheduling software, slurm. 
I tried multiple searches with different settings, and 10 raw files each: 
#1: Slurm setting: 1 node, 1 task/node, 1cpu/task + DIANN --thread 1 >> 95min DIA-NN processing time
#2: Slurm setting: 1 node, 24 task/node, 1cpu/task + DIANN --thread 24  >> 79min DIA-NN processing time
#3: Slurm setting: 2 node, 24 task/node, 1cpu/task + DIANN --thread 48  >> DIA-NN starts but crashes
#4: Slurm setting: 2 node, 24 task/node, 1cpu/task + DIANN --thread 24  >> 79min DIA-NN processing time

Q1: It appears that DIA-NN does not use the second node as it doesn't get faster. How to do that?
Q2: Looking at the log, DIA-NN doesn't parallelize the raw files in #2 or #4. It analyzes them sequentially as in #1. How is parallelization supposed to work in DIA-NN (if not analyzing raw files in parallel)?
Q3: I presume DIA-NN can be run on different nodes, correct? How to specify that?

**Comment by vdemichev** [comment, marked as answer]
> 1cpu/task + DIANN --thread 24

How can 24 threads run on a single CPU?

> DIA-NN starts but crashes

24 instances of DIA-NN, each requiring some RAM => probably physical RAM exhausted.

I would run 1 or 2 instances of DIA-NN per node, with the number of threads set to the number of physical cores.

**Reply by vdemichev** [reply]
Please see this thread https://github.com/bigbio/quantms/issues/164#issuecomment-1127600886 on our collaboration with PRIDE EBI on how to use DIA-NN to process files in a massively parallel fashion. Also, 1000 dia-PASEF runs might actually be fine to process on a single desktop, if powerful enough, in a reasonable time.

**Reply by vdemichev** [reply]
I am not familiar with 'clusters' and don't know what a 'node' is. You can run any number of instances of DIA-NN on any PC, provided there's enough RAM. Whether DIA-NN threads get assigned to run on particular cores - this is the choice of the operating system, DIA-NN just uses the standard C++11 multithreading API.

**Reply by JM-Bader** [reply]
> How can 24 threads run on a single CPU?
The cluster nodes have 24 cores (in a total of 1 or 2 CPUs), so 24 threads can be run per node.

> I am not familiar with 'clusters' and don't know what a 'node' is.
Clusters are set of computers that can work together - so they can be viewed as a single system.  A node refers to single computer/server of that cluster. 

> 24 instances of DIA-NN, each requiring some RAM => probably physical RAM exhausted.
We think there was a misunderstanding, sorry. We ran 24 instances of DIANN on one single node. That node has 24 physical cores. There was no issue. In that case physical RAM was not exhausted. 

A crash occured when specifying --thread 48 equivalent to the sum of 24 physical cores on node (computer) 1 and 24 physical cores on node (computer 2). 

----------------------------------------------------------------------------------
Fundamentally, our question is whether DIANN is able to process data on multiple computers (nodes) in parallel. From what we understand from your answer, DIANN multithreading is limited to multiple threads of one computer (as you normally use it) ?  Even if C++11 multithreading API supports multiple nodes (which we still need to clarify - or do you know?), probably one would need to specify such information with a DIANN command. 

Secondly, we don't really understand your example of processing 1000 dia-PASEF files in parallel on a single desktop (i.e computer). We are not aware of computers with 1000 cores (or at least we don't have one with that many cores or another way of running 1000 threads on a single computer).  Could there be a misunderstanding as well?

We read the thread on your collaboration with PRIDE EBI and didn't find the information on multithreading across computers or the number of threads on a single computer that they use. So unfortunately, it didn't solve our problem. 
Any help would be appreciated!

**Reply by vdemichev** [reply]
I am not sure what number of threads the PRIDE team sets, but I would recommed to run each DIA-NN instance with the number of threads equal to the number of physical cores of the CPU (here CPU means a chip on the motherboard) it is running on. Parallelism is achieved by processing different runs with different instances of DIA-NN.
1000 files one can process sequentially on a single desktop in a reasonable time.

**Reply by JM-Bader** [reply]
Thank your for your explanations and suggestions. We will find a way to process our large dataset. 

As an option for the future, we suggest to use HPX instead of the standard  C++11 multithreading API. This is compatible with SLURM which is widely used for parallelization on multiple nodes (as present in server clusters): 

https://hpx-docs.stellar-group.org/branches/master/html/index.html
https://hpx-docs.stellar-group.org/branches/master/html/manual/running_on_batch_systems.html#how-to-use-hpx-applications-with-slurm

This should further speed up searching of large raw data sets.

**Reply by vdemichev** [reply]
Thank you for the suggestion!
